### PR TITLE
JAMES-3881 Set a JMX password

### PR DESCRIPTION
**Pull request by vttranlina**
jira https://issues.apache.org/jira/browse/JAMES-3881

### CODE REVIEW

File: server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JMXServer.java [hunk 1]
```diff
@@ -126,4 +146,38 @@ private void doStop() {
         }
     }
 
+    private void generateJMXPasswordFileIfNeed() {
+        if (Boolean.parseBoolean(System.getProperty(JMX_CREDENTIAL_GENERATION_ENABLE_PROPERTY, JMX_CREDENTIAL_GENERATION_ENABLE_DEFAULT))
+            && !existJmxPasswordFile()) {
+            generateJMXPasswordFile();
+        }
+    }
```
Comment (chibenwa) [inline server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JMXServer.java:154]: Can we have the CLI attempt to read `~/conf/jmxremote.password` that is in clear if no option is supplied?

File: server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JMXServer.java [hunk 2]
```diff
@@ -126,4 +146,38 @@ private void doStop() {
...
+
+    private boolean existJmxPasswordFile() {
+        return Files.exists(Path.of(JmxConfiguration.PASSWORD_FILE_PATH)) && Files.exists(Path.of(JmxConfiguration.ACCESS_FILE_PATH));
+    }
+
+    private void generateJMXPasswordFile() {
+        File passwordFile = new File(JmxConfiguration.PASSWORD_FILE_PATH);
```
Comment (chibenwa) [inline server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JMXServer.java:161]: What are the permissions on this file?

File: server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JMXServer.java [hunk 3]
```diff
@@ -126,4 +146,38 @@ private void doStop() {
...
+            try (OutputStream outputStream = new FileOutputStream(passwordFile)) {
+                String randomPassword = RandomStringUtils.random(10, true, true);
+                IOUtils.write(JmxConfiguration.JAMES_ADMIN_USER_DEFAULT + " " + randomPassword + "\n", outputStream, StandardCharsets.UTF_8);
+                LOGGER.info("Generated JMX password file: " + JmxConfiguration.PASSWORD_FILE_PATH);
+            } catch (IOException e) {
+                throw new RuntimeException("Error when creating JMX password file: " + JmxConfiguration.PASSWORD_FILE_PATH, e);
+            }
+        }
+
+        File accessFile = new File(JmxConfiguration.ACCESS_FILE_PATH);
```
Comment (chibenwa) [inline server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JMXServer.java:172]: What are the permissions on this file?

File: server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JmxConfiguration.java [hunk 4]
```diff
@@ -34,6 +34,11 @@ public class JmxConfiguration {
     public static final String LOCALHOST = "localhost";
     public static final int DEFAULT_PORT = 9999;
     public static final boolean ENABLED = true;
+    public static final String JMX_CREDENTIAL_GENERATION_ENABLE_PROPERTY = "james.jmx.credential.generation";
+    public static final String JMX_CREDENTIAL_GENERATION_ENABLE_DEFAULT = "false";
+    public static final String PASSWORD_FILE_PATH = "/root/conf/jmxremote.password";
+    public static final String ACCESS_FILE_PATH = "/root/conf/jmxremote.access";
```
Comment (chibenwa) [inline server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JmxConfiguration.java:40]: This location is the docker one but we should not use that for raw installation. 

Suggestion:

```
org.apache.james.filesystem.api.FileSYstem.getFile("conf://jmxremote.password");
```

File: server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JmxConfiguration.java [hunk 5]
```diff
@@ -34,6 +34,11 @@ public class JmxConfiguration {
     public static final String LOCALHOST = "localhost";
     public static final int DEFAULT_PORT = 9999;
     public static final boolean ENABLED = true;
+    public static final String JMX_CREDENTIAL_GENERATION_ENABLE_PROPERTY = "james.jmx.credential.generation";
```
Comment (quantranhong1999) [inline server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JmxConfiguration.java:37]:
```suggestion
    public static final String JMX_CREDENTIAL_GENERATION_ENABLE_PROPERTY_KEY = "james.jmx.credential.generation";

File: server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JmxConfiguration.java [hunk 6]
```diff
@@ -34,6 +34,11 @@ public class JmxConfiguration {
...
+    public static final String JMX_CREDENTIAL_GENERATION_ENABLE_DEFAULT = "false";
```
Comment (quantranhong1999) [inline server/container/guice/jmx/src/main/java/org/apache/james/modules/server/JmxConfiguration.java:38]:
```suggestion
    public static final String JMX_CREDENTIAL_GENERATION_ENABLE_DEFAULT_VALUE = "false";
```
would be more readable to me

**Review by chibenwa — APPROVED** [review]

File: server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmx.adoc [hunk 7]
```diff
@@ -22,3 +22,37 @@ in GIT to get some examples and hints.
...
+
+- `jmxremote.access`: define the pair of username and access permission
+
+File's content example:
+```
+james-admin readWrite
+```
+
+When James runs with option `-Djames.jmx.credential.generation=true`, James will automatically generate `jmxremote.password` if the file does not exist.
+Then the default username is `james-admin` and a random password.
```
Comment (quantranhong1999) [inline server/apps/distributed-app/docs/modules/ROOT/pages/configure/jmx.adoc:47]: Could we document that by default the option `-Djames.jmx.credential.generation` is false and JMX authentication is off? And recommend operators turn it on instead.

If I do not read the code, I am not sure what happens if this option is false or I don't specify it, is it a breaking change or not...

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**